src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java

```diff
@@ -0,0 +1,396 @@
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static com.commercetools.sync.commons.MockUtils.getMockCustomerService;
+import static com.commercetools.sync.commons.MockUtils.getMockTypeService;
+import static com.commercetools.sync.commons.helpers.BaseReferenceResolver.BLANK_KEY_VALUE_ON_RESOURCE_IDENTIFIER;
+import static com.commercetools.sync.commons.helpers.CustomReferenceResolver.TYPE_DOES_NOT_EXIST;
+import static com.commercetools.sync.shoppinglists.helpers.ShoppingListReferenceResolver.CUSTOMER_DOES_NOT_EXIST;
+import static com.commercetools.sync.shoppinglists.helpers.ShoppingListReferenceResolver.FAILED_TO_RESOLVE_CUSTOMER_REFERENCE;
+import static com.commercetools.sync.shoppinglists.helpers.ShoppingListReferenceResolver.FAILED_TO_RESOLVE_CUSTOM_TYPE;
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class ShoppingListReferenceResolverTest {
+
+    private TypeService typeService;
+    private CustomerService customerService;
+
+    private ShoppingListReferenceResolver referenceResolver;
+
+    /**
+     * Sets up the services and the options needed for reference resolution.
+     */
+    @BeforeEach
+    void setup() {
+        typeService = getMockTypeService();
+        customerService = getMockCustomerService();
+
+        final ShoppingListSyncOptions syncOptions = ShoppingListSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+        referenceResolver = new ShoppingListReferenceResolver(syncOptions, customerService, typeService);
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithNonNullIdOnCustomTypeResId_ShouldResolveCustomTypeReference() {
+        final CustomFieldsDraft customFieldsDraft = CustomFieldsDraft
+            .ofTypeKeyAndJson("customTypeKey", new HashMap<>());
+
+        final ShoppingListDraftBuilder draftBuilder =
+            ShoppingListDraftBuilder
+                .of(LocalizedString.of(Locale.ENGLISH, "NAME"))
+                .custom(customFieldsDraft)
+                .description(LocalizedString.of(Locale.ENGLISH, "DESCRIPTION"));
+
+        // test
+        final ShoppingListDraftBuilder resolvedDraft = referenceResolver
+            .resolveCustomTypeReference(draftBuilder)
+            .toCompletableFuture().join();
+
+        // assertion
+        assertThat(resolvedDraft.getCustom()).isNotNull();
+        assertThat(resolvedDraft.getCustom().getType().getId()).isEqualTo("typeId");
+
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithNonNullKeyOnCustomTypeResId_ShouldResolveCustomTypeReference() {
```

Comment:
Should it be WithNonNullId for this test?

Reply:
check this name, please: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/e895c7a9806d27c3c07c712b42e3fbd894ff318c